0.9.x proactively create missing whisper files

adamvduke opened the pull request:
* extract whisper file creation to a function and move the function to storage.py
* extract recording in memory stats to a separate module to allow easily recording stats without circular dependencies
* move loading storage and aggregation schemas to storage.py
* proactively create whisper files for new metrics in a separate thread/service

If the metric is not in the cache AND the file for storing that metric does not exist, create the file to allow querying for those metrics immediately. This is helpful in the case that there is a large backlog of writing new metrics to disk which can cause a significant delay before being able to query for those metrics.

adamvduke committed:
extract whisper file creation to a function

the extracted function can then be used when uncached metrics come in, to proactively create the whisper file

* move whisper file creation to a function in storage.py
* move loading storage and aggregation schemas to storage.py

adamvduke committed:
proactively create whisper files for new metrics

If the metric is not in the cache AND the file for storing that metric does not exist, create the file to allow querying for those metrics immediately. This is helpful in the case that there is a large backlog of writing new metrics to disk which can cause a significant delay before being able to query for those metrics.

adamvduke committed:
separate recording internal stats from flushing to remove coupling

adamvduke committed:
ensure whisper file creates are instrumented in all cases

this ensures that whisper file creates are instrumented when they are caused by proactively creating the file and not only when they are caused by the writer loop

adamvduke committed:
add a settings config to allow turning on/off proactive whisper file creation

adamvduke committed:
create whisper files in a separate thread/service

When receiving metrics that are not currently in the cache, append a tuple with the metric name and database path, to a queue, which can be used in another thread/service to proactively create the whisper file. The new service continually tries to pop a tuple from the queue and create the whisper file. It is possible that both threads make it past the existence check for the file, but in that case, the writer assumes the file exists and attempts to update it anyway.

bmhatfield commented on lib/carbon/cache.py:
```diff
@@ -42,6 +45,10 @@ def gen_queue(self):
 
   def store(self, metric, datapoint):
     self.size += 1
+    dbFilePath = getFilesystemPath(metric)
```
This should go inside the `if settings.CACHE_CREATE_WHISPER_FILES` block.

adamvduke replied:
👍

bmhatfield commented on lib/carbon/cache.py:
```diff
@@ -42,6 +45,10 @@ def gen_queue(self):
     self.size += 1
+    dbFilePath = getFilesystemPath(metric)
+    if settings.CACHE_CREATE_WHISPER_FILES:
+      if (len(self[metric]) == 0):
```
I think you need to also check that the file does not exist.

adamvduke replied:
I intentionally left it out of the check to avoid having to query the file system here because I think this is a pretty hot path under load. We end up just bailing early later on if the file has already been created by the time we get to it.

bmhatfield commented on lib/carbon/cache.py:
```diff
@@ -42,6 +45,10 @@ def gen_queue(self):
+    if settings.CACHE_CREATE_WHISPER_FILES:
+      if (len(self[metric]) == 0):
+        queueWhisperCreate(metric, dbFilePath)
     self[metric].append(datapoint)
```
Maybe it makes sense to add this to the cache *only* if the file already exists, to avoid a race condition later? This suggestion feels iffy for some reason, though.

adamvduke replied:
I know the race is not ideal, but the current implementation rescues the exception and attempts to update the metric anyway. https://github.com/bmhatfield/carbon/blob/8d8d077ef739e6ad80b4c6d92d9a0dc9b19e0e36/lib/carbon/writer.py#L120-L131

It seems to me the worst case is that both threads do their existence check, get a `False` response and then one or the other will fail and log the exception. In the case that the writer rescues the exception, it logs the failure and attempts to update the metric anyway, and in the case that the whisper file creator rescues, it logs the exception and moves on.

bmhatfield replied:
It does not appear that this branch includes that exception handling code in the new `createWhisperFile` method.

adamvduke replied:
It allows `createWhisperFile` to raise, and then the caller rescues/logs.

https://github.com/bmhatfield/carbon/blob/e7840b3f86e2e10e3066dbf9a9531255ffb3030b/lib/carbon/writer.py#L90-L95

bmhatfield commented on conf/carbon.conf.example:
```diff
@@ -112,6 +112,11 @@ CACHE_QUERY_PORT = 7002
+# Set to True to enable the cache to proactively create whisper files
+# for metrics that are received but don't currently exist in the cache
+# and also do not already have an existing whisper file. Defaults to False.
+# CACHE_CREATE_WHISPER_FILES = True
```
Does it make sense to add an `_EARLY_` or change it to `PRECREATE` or something? Right now the name implies that turning this off would disable all creates, which is not what it does.

adamvduke replied:
Sure, I'm open to changing the name. I prefixed it with `CACHE` because it modifies/is used in the cache. Does that make sense?

`CACHE_PRECREATE_WHISPER_FILES`?

bmhatfield replied:
Sounds good.

bmhatfield reviewed (commented):
First pass comments/questions.

adamvduke committed:
calculate the whisper file system path when it's needed

rather than creating it up front and passing it through the queue, just create it before it's needed

adamvduke committed:
rename config option to turn on/off precreating whisper files

bmhatfield commented on lib/carbon/cache.py:
```diff
@@ -42,6 +45,9 @@ def gen_queue(self):
   def store(self, metric, datapoint):
     self.size += 1
+    if settings.CACHE_PRECREATE_WHISPER_FILES:
+      if (len(self[metric]) == 0):
```
Per our chat, let's switch this to `metric in self` as the test.

bmhatfield commented on lib/carbon/storage.py:
```diff
@@ -13,24 +13,84 @@
+    if not archiveConfig:
+      raise Exception("No storage schema matched the metric '%s', check your storage-schemas.conf file." % metric)
+
+    if not exists(dbFilePath):
```
Per our chat, please add an `else` to this with instrumentation that counts the amount of times we've hit this condition but the file already existed.

adamvduke committed:
optimize checking the cache for existing metrics

adamvduke committed:
instrument the number of times a whisper create call gets skipped
